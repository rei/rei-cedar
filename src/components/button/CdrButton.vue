<script setup>
import { useCssModule, computed, useSlots } from 'vue';

import mapClasses from '../../utils/mapClasses.js';
import { responsiveModifyClass, buildBooleanClass } from '../../utils/buildClass.js';
import propValidator from '../../utils/propValidator.js';

const slots = useSlots();
const props = defineProps({
tag: {
  type: String,
  default: 'button',
  validator: (value) => propValidator(value, ['button', 'a']),
},
type: {
  type: String,
  default: 'button',
  validator: (value) => propValidator(value, ['button', 'submit', 'reset']),
},
modifier: {
  type: String,
  default: 'primary',
  validator: (value) => propValidator(value, ['primary', 'secondary', 'sale', 'dark', 'link']),
},
size: {
  type: String,
  default: 'medium',
  validator: (value) => propValidator(
    value,
    ['small', 'medium', 'large'],
  ),
},
fullWidth: {
  type: [String, Boolean],
  default: false,
  validator: (value) => {
    if (typeof value === 'string') {
      return propValidator(
        value,
        ['@xs', '@sm', '@md', '@lg'],
        false,
      );
    }
    return typeof value === 'boolean';
  },
},
iconOnly: {
  type: Boolean,
  default: false,
},
withBackground: {
  type: Boolean,
  default: false,
},
});

const baseClass = 'cdr-button';
const style = useCssModule();

/************************  Computed properties *************************/
const buttonType = computed(() => props.tag === 'button' ? props.type : null);
const modifierClass = computed(() => `${baseClass}--${props.modifier}`);
const fullWidthClass = computed(() => !props.iconOnly && props.fullWidth
  && buildBooleanClass(baseClass, props.fullWidth, 'full-width'));
const sizeClass = computed(() => (!props.iconOnly
  ? responsiveModifyClass(baseClass, '', props.size)
  : `cdr-button--icon-only-${props.size}`));
const iconLeftClass = computed(() => slots['icon-left'] && slots.default
  && `${baseClass}--has-icon-left`);
const iconRightClass = computed(() => slots['icon-right'] && slots.default
  && `${baseClass}--has-icon-right`);
const iconOnlyClass = computed(() => props.iconOnly && `${baseClass}--icon-only`);
const withBackgroundClass = computed(() => props.iconOnly
  && props.withBackground && `${baseClass}--with-background`);

</script>
<template>
  <component
    :is="tag"
    :class="mapClasses(style,
                       baseClass,
                       modifierClass,
                       sizeClass,
                       fullWidthClass,
                       iconOnlyClass,
                       iconLeftClass,
                       iconRightClass,
                       iconOnlyClass,
                       withBackgroundClass,
    )"
    :type="buttonType"
  >
    <slot name="icon-left" />
    <slot />
    <slot name="icon-right" />
  </component>
</template>

<style lang="scss" module src="./styles/CdrButton.module.scss">
</style>
