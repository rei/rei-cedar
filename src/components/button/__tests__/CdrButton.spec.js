import { shallowMount, mount } from '../../../../test/vue-jest-style-workaround.js';
import CdrButton from '../CdrButton.vue';
import sinon from 'sinon'

describe('CdrButton', () => {
  describe('snapshot test', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = mount(CdrButton);
    })
    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  })

  describe('component unit tests', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = shallowMount(CdrButton);
    });

    describe('default component with no configuration', () => {
      it('renders correctly', () => {
        expect(wrapper.element).toMatchSnapshot();
      });
      it('sets the default type prop correctly', () => {
        expect(wrapper.attributes().type).toBe('button');
      });

      it('does not add icon class when slot is unused', () => {
        expect(wrapper.vm.iconClass).toBe(undefined);
      });
    })

    describe('with prop "type" set to "reset"', () => {
      beforeEach(() => {
        wrapper.setProps({ type: 'reset' });
      });
      it('renders correctly', () => {
        expect(wrapper.element).toMatchSnapshot();
      });
      it('sets type attr correctly', () => {
        expect(wrapper.attributes().type).toBe('reset');
      });
    })

    describe('when "el" is set to "a"', () => {
      beforeEach(() => {
        wrapper.setProps({ el: 'a' });
      });
      it('renders correctly', () => {
        expect(wrapper.element).toMatchSnapshot();
      });
      it('computes base class correctly', () => {
        expect(wrapper.classes()).toContain('cdr-button');
      });
    })

    describe('when "iconOnly" is set to true and size is "large"', () => {
      beforeEach(() => {
        wrapper.setProps({
          size: 'large',
          iconOnly: true,
        });
      });
      it('renders correctly', () => {
        expect(wrapper.element).toMatchSnapshot();
      });
      it('adds custom size class', () => {
        expect(wrapper.vm.sizeClass).toBe('cdr-button--icon-only-large');
      });
    })

    describe('when "iconOnly" and "withBackground" is set to true', () => {
      beforeEach(() => {
        wrapper.setProps({
          iconOnly: true,
          withBackground: true,
        });
      });
      it('renders correctly', () => {
        expect(wrapper.element).toMatchSnapshot();
      });
      it('adds class for icon only', () => {
        expect(wrapper.classes()).toContain('cdr-button--icon-only');
      });

      it('adds class for icon background', () => {
        expect(wrapper.classes()).toContain('cdr-button--with-background');
      });
    })

    describe('with tag prop set to "a"', () => {
      beforeEach(() => {
        wrapper.setProps({ tag: 'a' });
      });
      it('renders correctly', () => {
        expect(wrapper.element).toMatchSnapshot();
      });
      it('renders a link', () => {
        expect(wrapper.element.tagName).toBe('A');
      });
    })

    describe('with a function passed to the onClick attribute', () => {
      let spy;
      beforeEach(() => {
        spy = sinon.spy();
        wrapper = shallowMount(CdrButton, {
          attrs: {
            onClick: spy
          },
        });
        wrapper.trigger('click');
      });
      it('renders correctly', () => {
        expect(wrapper.element).toMatchSnapshot();
      });
      it('the function is called', () => {
        expect(spy.calledOnce).toBeTruthy();
      });
    })



    describe('with an icon-left slot', () => {
      beforeEach(() => {
        wrapper = shallowMount(CdrButton, {
          slots: {
            default: ['default slot for text'],
            'icon-left': '<div />',
          },
        });
      });
      it('renders correctly', () => {
        expect(wrapper.element).toMatchSnapshot();
      });
      it('adds the "cdr-button--has-icon-left" class', ()=>{
        expect(wrapper.classes()).toContain('cdr-button--has-icon-left');
      })
    });

    describe('with an icon-left slot', () => {
      beforeEach(() => {
        wrapper = shallowMount(CdrButton, {
          slots: {
            default: ['default slot for text'],
            'icon-right': '<div />',
          },
        });
      });
      it('renders correctly', () => {
        expect(wrapper.element).toMatchSnapshot();
      });
      it('adds the "cdr-button--has-icon-right" class', ()=>{
        expect(wrapper.classes()).toContain('cdr-button--has-icon-right');
      })
    });
  });
});