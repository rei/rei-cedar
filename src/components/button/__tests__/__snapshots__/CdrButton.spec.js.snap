// Vitest Snapshot v1

exports[`CdrButton > renders correctly 1`] = `
<button
  class="cdr-button cdr-button--primary cdr-button--medium"
  type="button"
>
  
  
  
  
  
  
</button>
`;
